File: androidApp/src/main/java/jp/hotdrop/considercline/android/ui/pointget/PointGetViewModel.kt
package jp.hotdrop.considercline.android.ui.pointget

import dagger.hilt.android.lifecycle.HiltViewModel
import jp.hotdrop.considercline.android.ui.BaseViewModel
import jp.hotdrop.considercline.di.KmpFactory
import jp.hotdrop.considercline.model.AppComplete
import jp.hotdrop.considercline.model.AppError
import jp.hotdrop.considercline.model.AppResult
import jp.hotdrop.considercline.model.Point
import jp.hotdrop.considercline.usecase.PointUseCase
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class PointGetViewModel @Inject constructor() : BaseViewModel() {
    private val pointUseCase: PointUseCase by lazy { KmpFactory.useCaseFactory.pointUseCase }

    private val _uiState = MutableStateFlow(PointGetUiState())
    val uiState: StateFlow<PointGetUiState> = _uiState.asStateFlow()

    init {
        launch {
            _uiState.update { it.copy(isStartScreenLoading = true) }
            when(val result = pointUseCase.find()) {
                is AppResult.Success -> _uiState.update {
                    it.copy(currentPoint = result.data, isStartScreenLoading = false)
                }
                is AppResult.Error -> _uiState.update {
                    it.copy(loadingErrorMessage = result.error.message, isStartScreenLoading = false)
                }
            }
        }
    }

    fun inputPoint(newInputPoint: Int, maxAvailablePoint: Int) {
        val errorMessage = validateInput(
            input = newInputPoint,
            current = _uiState.value.currentPoint.balance,
            maxAvailablePoint = maxAvailablePoint
        )
        _uiState.update {
            it.copy(
                inputPoint = newInputPoint,
                inputPointErrorMessage = errorMessage,
                isEnableInputPoint = errorMessage == null && newInputPoint > 0
            )
        }
    }

    fun acquirePoint(inputPoint: Int) {
        launch {
            _uiState.update { it.copy(runAcquiringProcess = true) }
            when (val result = pointUseCase.acquire(inputPoint)) {
                AppComplete.Complete -> _uiState.update {
                    it.copy(acquireEvent = PointAcquireEvent.ShowSuccessDialog, runAcquiringProcess = false)
                }
                is AppComplete.Error -> _uiState.update {
                    it.copy(acquireEvent = PointAcquireEvent.ShowErrorDialog(result.error), runAcquiringProcess = false)
                }
            }
        }
    }

    fun resetAcquireEvent() {
        _uiState.update { it.copy(acquireEvent = null) }
    }

    private fun validateInput(input: Int, current: Int, maxAvailablePoint: Int): String? {
        return when {
            input <= 0 -> "0より大きい値を入力してください。"
            input + current > maxAvailablePoint -> "最大ポイントを超えています。"
            else -> null
        }
    }
}

data class PointGetUiState(
    val currentPoint: Point = Point(0),
    val inputPoint: Int = 0,
    val isStartScreenLoading: Boolean = false,
    val loadingErrorMessage: String? = null,
    val inputPointErrorMessage: String? = null,
    val isEnableInputPoint: Boolean = false,
    val acquireEvent: PointAcquireEvent? = null,
    val runAcquiringProcess: Boolean = false
)

sealed class PointAcquireEvent {
    object ShowSuccessDialog: PointAcquireEvent()
    data class ShowErrorDialog(val error: AppError): PointAcquireEvent()
}

File: iosApp/iosApp/Sources/ui/pointget/PointGetViewModel.swift

import Foundation
import shared

class PointGetViewModel: ObservableObject {
    @Published var viewState: PointGetViewState = .loading
    @Published var acquireEventState: PointAcquireEventState?

    private let pointUseCase: PointUseCase
    // TODO　本来は設定ファイル（config？）から取得する
    private let maxPoint = 20000

    init(pointUseCase: PointUseCase = KmpFactory.shared.useCaseFactory.pointUseCase) {
        self.pointUseCase = pointUseCase
    }

    @MainActor
    func load() async {
        do {
            let point = try await pointUseCase.findForIos()
            viewState = .success(
                currentPoint: point,
                inputPoint: 0,
                errorMessage: nil,
                isEnableConfirm: false
            )
        } catch {
            viewState = .error(message: "ポイント残高の取得に失敗しました。")
        }
    }

    func inputPoint(newInputPoint: Int) {
        guard case let .success(currentPoint, _, _, _) = viewState else {
            return
        }

        let errorMessage = validateInput(
            input: newInputPoint,
            current: Int(currentPoint.balance),
            maxAvailablePoint: maxPoint
        )
        let isEnable = errorMessage == nil && newInputPoint > 0

        viewState = .success(
            currentPoint: currentPoint,
            inputPoint: newInputPoint,
            errorMessage: errorMessage,
            isEnableConfirm: isEnable
        )
    }

    @MainActor
    func acquirePoint(inputPoint: Int) async {
        do {
            try await pointUseCase.acquire(inputPoint: Int32(inputPoint))
            acquireEventState = .success
        } catch {
            acquireEventState = .error(message: "ポイントの獲得に失敗しました。")
        }
    }

    private func validateInput(input: Int, current: Int, maxAvailablePoint: Int) -> String? {
        if input <= 0 {
            return "0より大きい値を入力してください。"
        }
        if input + current > maxAvailablePoint {
            return "最大ポイントを超えています。"
        }
        return nil
    }
}

enum PointGetViewState: Equatable {
    case loading
    case success(currentPoint: Point, inputPoint: Int, errorMessage: String?, isEnableConfirm: Bool)
    case error(message: String)
}

enum PointAcquireEventState: Equatable {
    case success
    case error(message: String)
}

